src/frontend/query_commands.ml

```diff
@@ -359,6 +359,43 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
     in
     List.map ~f:Mbrowse.node_loc path
 
+  | Locate_type pos ->
+    let typer = Mpipeline.typer_result pipeline in
+    let structures = Mbrowse.of_typedtree (Mtyper.get_typedtree typer) in
+    let pos = Mpipeline.get_lexing_pos pipeline pos in
+    let path = Mbrowse.enclosing pos [structures] in
+    let path =
+      Std.List.filter_map path ~f:(fun (env, node) ->
```

Comment:
This looks weird to me. IIUC you are call locate on the type of all the enclosing nodes … is that really useful / what you want?

Admittedly the API to get the deepest node around the cursor is … nonexistent (it's something that we should probably improve), but you can get inspiration from the way `Case_analysis` is handled lower down in this file.

Reply:
Indeed, it's a bit odd. That's how it was originally.

@andreypopp Was there a reason you looked up the location for the type of all enclosing nodes?

Reply:
> Admittedly the API to get the deepest node around the cursor is … nonexistent (it's something that we should probably improve), but you can get inspiration from the way Case_analysis is handled lower down in this file.

Isn't the deepest node around the cursor just:
```
List.hd (Mbrowse.enclosing cursor [structures]
```

I've looked into `Case_analysis`, but from what I see the filtering there is necessary to find a node that spans the user's selection. 

So it seems like I could just `find_map` instead of `filter_map`?

Reply:
IIRC the intention was to make the query fallback to locating the outer node if (for some reason) it cannot do this for the one under the cursor. Now I don't think this is useful.

Though my thinking back then was that it can be useful if we'd return the range of the node we've located the type for, so we can highlight it in an editor (I think vscode can preview go-to-definition location in a popup and thus you can see both the original and located positions at the same time).

Reply:
Yeah, that seems like a good thing to have, but I think that the current implementation has some surprising side effects.

For example, consider this fragment from query_commands.ml (around line 629):

```ocaml
    let nodes =
      List.drop_while nodes
        ~f:(fun (_,t) ->
          let {Location. loc_start; loc_end; _} = Mbrowse.node_loc t in
                                        (* cursor on the [pos_start] below *)
          Lexing.compare_pos loc_start pos_start > 0 || Lexing.compare_pos loc_end pos_end < 0)
```

Attempting to locate the type will go to definition of list in `listLabels.mli`. That seems very confusing to me

Reply:
> Isn't the deepest node around the cursor just: `List.hd (Mbrowse.enclosing cursor [structures])`

Probably :)